lisa/sut_orchestrator/azure/features.py

```diff
@@ -108,10 +121,211 @@ def _execute(self, wait: bool, operator: str, **kwargs: Any) -> None:
+        retry = 0
+        for retry in range(3):
+            # websocket connection is not stable, so we need to retry
+            try:
+                # run command with timeout
+                self._asyncio_loop.run_until_complete(
+                    asyncio.wait_for(
+                        self._write_async(cmd),
+                        timeout=timeout,
+                    )
+                )
+                return
+            except websockets.ConnectionClosed as e:  # type: ignore
+                # If the connection is closed, we need to reconnect
+                self._log.debug(f"Connection closed: {e}")
+                self._ws = None
+                self._asyncio_loop.run_until_complete(
+                    asyncio.wait_for(
+                        self._get_connection_async(),
+                        timeout=timeout,
+                    )
+                )
+                retry += 1
+                continue
+            except Exception as e:
+                self._log.debug(f"Failed to run command: {e}")
+                retry += 1
+                continue
+
+        raise LisaException("Failed to write command to serial console")
+
+    def read(self, timeout: int = 120) -> str:
+        retry = 0
+        for retry in range(3):
+            # websocket connection is not stable, so we need to retry
+            try:
+                # run command with timeout
+                msg = self._asyncio_loop.run_until_complete(
+                    asyncio.wait_for(
+                        self._read_async(),
+                        timeout=timeout,
+                    )
+                )
+                return msg
+            except websockets.ConnectionClosed as e:  # type: ignore
+                # If the connection is closed, we need to reconnect
+                self._log.debug(f"Connection closed: {e}")
+                self._ws = None
+                self._asyncio_loop.run_until_complete(
+                    asyncio.wait_for(
+                        self._get_connection_async(),
+                        timeout=timeout,
+                    )
+                )
+                retry += 1
+                continue
+            except Exception as e:
+                self._log.debug(f"Failed to run command: {e}")
+                retry += 1
+                continue
+
+        raise LisaException("Failed to read from serial console")
+
+    async def _get_connection_async(self) -> Any:
+        if self._ws is None:
+            self._log.debug("Creating connection to serial console")
+            connection_str = self._get_connection_string()
+            self._ws = await websockets.connect(connection_str)  # type: ignore
+
+        return self._ws
+
+    async def _write_async(self, cmd: str) -> None:
+        # connect to websocket and send command
+        ws = await self._get_connection_async()
+        await ws.send(cmd)
+
+    async def _read_async(self) -> str:
```

Comment:
Because the async method is not to expose, so it doesn't need to use async here. The self._asyncio_loop should be in the method, and make those method synchronized.

Reply:
I didn't understand this comment. Do you mean remove all async functions?

Reply:
See when the method it's called, there is a "_asyncio_loop" always. It means the method doesn't need to be "async", The async should be wrapped in the method itself. If the websocket support sync calls, it should be used. Because in LISA, we don't use the async/await.

Reply:
Got it! I think saw some sync examples with websocket api. Let me check those.